File: resources/php/forgot2.php
<?php

session_start();

date_default_timezone_set("Asia/Kolkata");

require __DIR__ . '/mail/Exception.php';
require __DIR__ . '/mail/PHPMailer.php';
require __DIR__ . '/mail/SMTP.php';

use PHPMailer\PHPMailer\Exception;
use PHPMailer\PHPMailer\PHPMailer;

include 'config.php';
include 'smtp1.php';

try {
    // Create connection
    $conn = new mysqli($servername, $username, $password, $dbname);

    // Check connection
    if ($conn->connect_error) {
        throw new Exception("Connection failed: " . $conn->connect_error);
    }

    // Check if the form is submitted
    if ($_SERVER["REQUEST_METHOD"] == "POST") {

        if (isset($_POST["email"])) {
            // Retrieve form data
            $emailpt = htmlspecialchars(trim($_POST["email"]));
                
            $checkQuery = "SELECT adminid, username, email FROM admin WHERE email = ?";

            if (!($checkStmt = $conn->prepare($checkQuery))) {
                throw new Exception("Prepare failed: (" . $conn->errno . ") " . $conn->error);
            }

            if (!$checkStmt->bind_param("s", $emailpt)) {
                throw new Exception("Binding parameters failed: (" . $checkStmt->errno . ") " . $checkStmt->error);
            }

            if (!$checkStmt->execute()) {
                throw new Exception("Execute failed: (" . $checkStmt->errno . ") " . $checkStmt->error);
            }

            $checkStmt->store_result();
            $checkStmt->bind_result($adminid, $username, $email); 
            $checkStmt->fetch();

            if ($checkStmt->num_rows() > 0) {
                if ($emailpt == $email) {

                    $_SESSION['adminid'] = $adminid;
                    $_SESSION['email'] = $email;

                    $otppt = str_pad(random_int(0, 999999), 6, '0', STR_PAD_LEFT);

                    $mail = new PHPMailer(true);

                    try {
                        //Server settings
                        $mail->SMTPDebug = 0;                                 
                        $mail->isSMTP();                                      
                        $mail->Host = $smtphost;  
                        $mail->SMTPAuth = true;                               
                        $mail->Username = $smtpusername;                 
                        $mail->Password = $smtppassword;                           
                        $mail->SMTPSecure = PHPMailer::ENCRYPTION_STARTTLS;                            
                        $mail->Port = $smtpport;                                    
            
                        //Recipients
                        $mail->setFrom($smtpusername, 'Foodelight');
                        $mail->addAddress($emailpt);     
            
                        //Content
                        $mail->isHTML(true);                                  
                        $mail->Subject = 'Reset Password';
                        $mail->Body    = "Hi " . $username . ",<br><br>Your OTP for password reset is: " . $otppt . "<br><br>This OTP is valid for 2 minutes only.<br><br>Please use this OTP to reset your password.<br><br>Thanks,<br>Foodelight";
                        
                        $result = $mail->send();
                        if ($result) {
                            echo 'OTP has been sent to your email';

                            $expiresAt = date('Y-m-d H:i:s', time() + 120);
                            $isUsed = 0;
                            $checkQuery1 = "INSERT INTO admin_otp(adminid, otp, expires_at, is_used) VALUES (?, ?, ?, ?)";

                            if (!($checkStmt1 = $conn->prepare($checkQuery1))) {
                                throw new Exception("Prepare failed: (" . $conn->errno . ") " . $conn->error);
                            }

                            if (!$checkStmt1->bind_param("iisi", $adminid, $otppt, $expiresAt, $isUsed)) {
                                throw new Exception("Binding parameters failed: (" . $checkStmt1->errno . ") " . $checkStmt1->error);
                            }

                            if (!$checkStmt1->execute()) {
                                throw new Exception("Error: " . $checkStmt1->error);
                            }

                            $checkStmt1->close();

                        } else {
                            echo "ERROR";
                        }

                    } catch (Exception $e) {
                        echo 'Message could not be sent. Mailer Error: ', $mail->ErrorInfo;
                    }
                } else {
                    echo "Invalid credentials1";
                }

            } else {
                echo "Account not Found!";
            }
            $checkStmt->close();
        } else if (isset($_POST["otp"])) {
            try {
                $adminid = $_SESSION['adminid'] ?? null;
                $otp = $_POST["otp"];
                if (!$adminid) {
                    throw new Exception("Session expired or invalid. Please try again.");
                }
                $checkStmt = $conn->prepare("SELECT id FROM admin_otp WHERE adminid = ? AND otp = ? AND is_used = 0 AND NOW() <= expires_at");
                if ($checkStmt === false) {
                    throw new Exception($conn->error);
                }
                $checkStmt->bind_param("ii", $adminid, $otp);
                $checkStmt->execute();
                $result = $checkStmt->get_result();
            
                if ($result->num_rows > 0) {
                    $otpRow = $result->fetch_assoc();
                    $otp_id = $otpRow['id'];

                    $updateStmt = $conn->prepare("UPDATE admin_otp SET is_used = 1 WHERE id = ?");
                    if ($updateStmt === false) {
                        throw new Exception($conn->error);
                    }
                    $updateStmt->bind_param("i", $otp_id);
                    $updateStmt->execute();
                    $updateStmt->close();

                    echo "OTP is valid.";
                } else {
                    echo "Invalid OTP!";
                }
                $checkStmt->close();
            } catch (Exception $e) {
                die("Error: " . $e->getMessage());
            }
        } else {
            // Handle case where one or more keys are not set
            throw new Exception("Error: One or more form fields are missing.");
        }

    }

    // Close the database connection
    $conn->close();
} catch (Exception $e) {
    echo  $e->getMessage(), "\n";
}
?>

File: python/forgot2.py
import random
from datetime import timedelta

from flask import request, session

from common import clean
from common import is_expired_pst_value, now_pst_naive
from db import get_connection
from db_init import initialize_schema
from mailer import send_html_mail


def register_routes(app):
    @app.route("/python/forgot2.py", methods=["POST"])
    def forgot2_php():
        conn = None
        try:
            conn = get_connection()
            initialize_schema(conn)

            if request.method == "POST":
                form = request.form

                if "email" in form:
                    emailpt = clean(form.get("email", ""))
                    with conn.cursor() as cur:
                        cur.execute("SELECT adminid, username, email FROM admin WHERE email = %s", (emailpt,))
                        row = cur.fetchone()

                    if row:
                        adminid = row.get("adminid")
                        username = row.get("username", "")
                        email = row.get("email", "")
                        session["adminid"] = adminid
                        session["username"] = username
                        session["email"] = email

                        otppt = str(random.randint(0, 999999)).zfill(6)
                        body = (
                            "Hi "
                            + username
                            + ",<br><br>Your OTP for password reset is: "
                            + otppt
                            + "<br><br>This OTP is valid for 2 minutes only.<br><br>Please use this OTP to reset your password.<br><br>Thanks,<br>Foodelight"
                        )
                        sent, error = send_html_mail(emailpt, "Reset Password", body)
                        if sent:
                            expires_at = now_pst_naive() + timedelta(minutes=2)
                            with conn.cursor() as cur:
                                cur.execute("UPDATE admin_otp SET is_used = TRUE WHERE adminid = %s AND is_used = FALSE", (adminid,))
                                cur.execute(
                                    "INSERT INTO admin_otp(adminid, otp, expires_at, is_used) VALUES (%s, %s, %s, %s)",
                                    (adminid, otppt, expires_at, False),
                                )
                            conn.commit()
                            return "OTP has been sent to your email"
                        return "ERROR" if not error else "Message could not be sent. Mailer Error: " + error

                    return "Account not Found!"

                if "resend_admin_forgot_otp" in form:
                    adminid = session.get("adminid")
                    username = session.get("username", "")
                    email = session.get("email", "")

                    if not adminid or not email:
                        return "Session expired or invalid. Please try again."

                    otppt = str(random.randint(0, 999999)).zfill(6)
                    body = (
                        "Hi "
                        + str(username)
                        + ",<br><br>Your OTP for password reset is: "
                        + otppt
                        + "<br><br>This OTP is valid for 2 minutes only.<br><br>Please use this OTP to reset your password.<br><br>Thanks,<br>Foodelight"
                    )
                    sent, error = send_html_mail(str(email), "Reset Password", body)
                    if sent:
                        expires_at = now_pst_naive() + timedelta(minutes=2)
                        with conn.cursor() as cur:
                            cur.execute("UPDATE admin_otp SET is_used = TRUE WHERE adminid = %s AND is_used = FALSE", (adminid,))
                            cur.execute(
                                "INSERT INTO admin_otp(adminid, otp, expires_at, is_used) VALUES (%s, %s, %s, %s)",
                                (adminid, otppt, expires_at, False),
                            )
                        conn.commit()
                        return "OTP has been sent to your email"
                    return "ERROR" if not error else "Message could not be sent. Mailer Error: " + error

                if "otp" in form:
                    adminid = session.get("adminid")
                    otp = (form.get("otp", "") or "").strip()
                    if not adminid:
                        raise Exception("Session expired or invalid. Please try again.")
                    if not otp:
                        return "Invalid OTP!"

                    with conn.cursor() as cur:
                        cur.execute(
                            "SELECT id, expires_at FROM admin_otp WHERE adminid = %s AND otp = %s AND is_used = FALSE",
                            (adminid, otp),
                        )
                        row = cur.fetchone()

                    if row:
                        expires_at = row.get("expires_at")
                        if expires_at is None:
                            return "Invalid OTP!"
                        if is_expired_pst_value(expires_at):
                            return "OTP expired!"

                        with conn.cursor() as cur:
                            cur.execute("UPDATE admin_otp SET is_used = TRUE WHERE id = %s", (row.get("id"),))
                        conn.commit()
                        return "OTP is valid."

                    return "Invalid OTP!"

                raise Exception("Error: One or more form fields are missing.")

            return ""
        except Exception as exc:
            return "Error: " + str(exc)
        finally:
            if conn is not None:
                conn.close()

